types/react-side-effect/index.d.ts

```diff
@@ -1,22 +1,21 @@
-// Type definitions for react-side-effect v1.0.2
+// Type definitions for react-side-effect 1.0
 // Project: https://github.com/gaearon/react-side-effect
 // Definitions by: Remo H. Jansen <https://github.com/remojansen>
+//                 Martin Charles <https://github.com/0xcaff>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.6
 
 import * as React from "react";
 
-declare function withSideEffect(
-    reducePropsToState: (propsList: any[]) => any,
-    handleStateChangeOnClient: (state: any) => any,
-    mapStateOnServer?: (state: any) => any
-): ClassDecorator;
+declare function withSideEffect<TProp, TState>(
+    reducePropsToState: (propsList: TProp[]) => TState,
+    handleStateChangeOnClient: (state: TState) => void,
+    mapStateOnServer?: (state: TState) => void
+): ClassDecorator<TProp>;
 
-declare class ElementClass extends React.Component<any> {}
-
-interface ClassDecorator {
-    <T extends (typeof ElementClass)>(component:T): T;
-}
+type ClassDecorator<TProp> = (
+    component: React.ComponentType<TProp>
+) => React.ComponentType<TProp>;
```

Comment:
This should return `React.ComponentType<TProp> & {peek(): TState | TServerState, rewind(): TServerState}`